python/src/executor.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Another thread that can run async tasks in the background.
+
+// See: https://thenewstack.io/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/
+
+use std::sync::{Arc, Mutex};
+
+use futures::{future::BoxFuture, Future};
+use pyo3::Python;
+
+/// A background executor which allows running tasks on a tokio runtime
+/// in a separate thread.
+pub struct BackgroundExecutor {
+    state: Arc<Mutex<State>>,
+}
+
+struct State {
+    /// Channel for requests. The executor takes futures off this channel
+    /// and spawns them on the runtime.                                                                                                    
+    requests: Option<std::sync::mpsc::Sender<BoxFuture<'static, ()>>>,
+
+    /// Thread which contains a tokio runtime.
+    _thread: std::thread::JoinHandle<()>,
+}
+
+impl BackgroundExecutor {
```

Comment:
Why is this background executor needed?  It feels almost like we have a proprietary runtime guarding access to the tokio runtime.  In other words, why doesn't `RT.block_on` do something like `self.tokio_runtime.block_on(task)`?

Also, performance-wise won't this prevent any parallel usage of lance?

Reply:
In tokio, `Runtime.block_on()` runs the future *in the current thread*. Importantly, if it is called from within a async tokio runtime, tokio will panic. This happens now if for example you try to pass one dataset as input to another:

```python
existing_dataset = lance.dataset('path')
# Creates an iterator that calls `block_on()` on each iteration:
reader = dataset.to_batches()
# Calls `block_on()` to consumer the iterator and write data
lance.write_dataset(existing_dataset, 'new_path', schema=existing_dataset.schema) # Panics!
```

> Also, performance-wise won't this prevent any parallel usage of lance?

It shouldn't. There's nothing to stop multiple Python threads from calling into `RT.block_on()`.

Reply:
I suppose if we wanted, we could only use the background executor for things that might be called from Rust in a nested fashion. The background executor could store the runtime in an `Arc<>` and we could keep a copy of it on the main thread so it can be accessed directly for functions that can use the runtime directly. What do you think of that?

```rust
pub struct BackgroundExecutor {
    state: Arc<Mutex<State>>,
    pub runtime: Arc<tokio::runtime::Runtime>,
}
```

Reply:
But to answer the more fundamental question, "why do we need a background executor?": There are cases where we don't want the current thread to be the one that is polling the futures, but it has to be polled by someone, so we submit it to the background thread.

Reply:
I see.  I was misreading this.  I thought the background executor was doing something like...

```
while True:
  next_task = rx.recv()
  result = runtime.block_on(next_task)
  tx.send(result)
```

I see now that all the background executor is doing is submitting jobs to the shared runtime.  But why does this job submission need to be on a separate thread?  Couldn't you just call `spawn` from the python thread?

> But to answer the more fundamental question, "why do we need a background executor?": There are cases where we don't want the current thread to be the one that is polling the futures, but it has to be polled by someone, so we submit it to the background thread.

Doesn't the polling actually happen here (which IS on the current thread):

>  rx.recv().unwrap()

Reply:
> But why does this job submission need to be on a separate thread? Couldn't you just call spawn from the python thread?

Initially, I thought it might not work if theres not at least one thread calling `block_on`. But apparently it does.

I've reworked the executor design to spawn directly onto the runtime.